File: src/main/python/lab999_functions/zipWithApp.py
"""
 zip_with function.

 @author rambabu.posa
"""
from pyspark.sql import (SparkSession, Row)
from pyspark.sql.types import (StructType, StructField, IntegerType, ArrayType)
from pyspark.sql import (functions as F)
import os

current_dir = os.path.dirname(__file__)
relative_path = "../../../../data/functions/functions.csv"
absolute_file_path = os.path.join(current_dir, relative_path)

# Creates a session on a local master
spark = SparkSession.builder.appName("zip_with function") \
    .master("local[*]").getOrCreate()


def create_dataFrame(spark):
    schema = StructType([StructField("c1", ArrayType(IntegerType()), False),
                         StructField("c2", ArrayType(IntegerType()), False)])
    df = spark.createDataFrame([Row(c1=[1010, 1012], c2=[1021, 1023, 1025]),
                                Row(c1=[2010, 2012, 2014], c2=[2021, 2023]),
                                Row(c1=[3010, 3012], c2=[3021, 3023])
                                ], schema)
    return df


df = df = create_dataFrame(spark)

print("Input")
df.show(5, False)

df.printSchema()

def myfun(c1,c2):
    return c1+c2

#df = df.withColumn("zip_with", F.arrays_zip(F.col("c1"), F.col("c2")))
#df = df.withColumn("zipped_add", F.expr("transform(zip_with, x -> (x.c1 + x.c2))"))
df = df.withColumn("zip_with", F.expr("zip_with(c1, c2, (x,y) -> (x+y))"))
#df = df.withColumn("zipped_final2", F.expr("zip_with(c1, c2, (x,y) -> (x+y)"))
#df = df.select("c1,c2,zip_with(c1,c2, (x,y) -> sum(x,y))")
#df = df.withColumn("zip_with", F.expr("transform(zip_with, x -> when(x.c1.isNull() | x.c2.isNull(), lit(-1)).otherwise(x.c1 + x.c2))"))


print("After zip_with")
df.show(5, False)
df.printSchema()

spark.stop()

File: src/main/java/net/jgp/books/spark/ch13/lab999_functions/ZipWithApp.java
package net.jgp.books.spark.ch13.lab999_functions;

import static org.apache.spark.sql.functions.col;
import static org.apache.spark.sql.functions.lit;
import static org.apache.spark.sql.functions.when;
import static org.apache.spark.sql.functions.zip_with;

import java.util.ArrayList;
import java.util.List;

import org.apache.spark.sql.Column;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.RowFactory;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.types.DataTypes;
import org.apache.spark.sql.types.StructField;
import org.apache.spark.sql.types.StructType;

import scala.Function2;

/**
 * zip_with function.
 * 
 * @author jgp
 */
public class ZipWithApp {
  private SparkSession spark = null;

  /**
   * Takes two values and adds them. If one of the value is null, as the
   * addition cannot be performed, the result is -1.
   * 
   * The processing is taking place in the apply() method of the function.
   * It can be implemented as a Java lambda function.
   * 
   * @author jgp
   */
  public class ZipWithFunction
      implements Function2<Column, Column, Column> {

    @Override
    public Column apply(Column v1, Column v2) {
      return when(v1.isNull().or(v2.isNull()), lit(-1))
          .otherwise(v1.$plus(v2));
    }

  }

  public static void main(String[] args) {
    ZipWithApp app = new ZipWithApp();
    app.start();
  }

  /**
   * The processing code.
   */
  private void start() {
    spark = SparkSession.builder()
        .appName("abs function")
        .master("local[*]")
        .getOrCreate();

    Dataset<Row> df = createDataframe();
    System.out.println("Input");
    df.show(5, false);

    df = df.withColumn("zip_with",
        zip_with(col("c1"), col("c2"), new ZipWithFunction()));

    System.out.println("After zip_with");
    df.show(5, false);
  }

  /**
   * Creates a dataframe containing arrays of integers.
   */
  private Dataset<Row> createDataframe() {
    StructType schema = DataTypes.createStructType(new StructField[] {
        DataTypes.createStructField(
            "c1",
            DataTypes.createArrayType(DataTypes.IntegerType),
            false),
        DataTypes.createStructField(
            "c2",
            DataTypes.createArrayType(DataTypes.IntegerType),
            false) });

    List<Row> rows = new ArrayList<>();
    rows.add(RowFactory.create(
        new int[] { 1010, 1012 },
        new int[] { 1021, 1023, 1025 }));
    rows.add(RowFactory.create(
        new int[] { 2010, 2012, 2014 },
        new int[] { 2021, 2023 }));
    rows.add(RowFactory.create(
        new int[] { 3010, 3012 },
        new int[] { 3021, 3023 }));

    return spark.createDataFrame(rows, schema);
  }
}
